File: src/backend/reservation/src/main/java/org/taonaw/reservation/domain/model/cancellationfeesetting/CancellationFeeSetting.java
package org.taonaw.reservation.domain.model.cancellationfeesetting;

import lombok.AllArgsConstructor;
import lombok.NonNull;
import org.taonaw.reservation.domain.model.reservation.UseTime;

import java.time.LocalDate;
import java.util.Set;

@AllArgsConstructor
public class CancellationFeeSetting {
    @NonNull
    private final Set<CancellationFeeRate> cancellationFeeRates;

    public boolean isFree(@NonNull UseTime useTime, @NonNull LocalDate currentDate) {
        for (var cancellationFeeRate : cancellationFeeRates) {
            if (cancellationFeeRate.isApplied(useTime, currentDate) && cancellationFeeRate.isNotFree()) {
                return false;
            }
        }
        return true;
    }

    public boolean isNotFree(@NonNull UseTime useTime, @NonNull LocalDate currentDate) {
        return !isFree(useTime, currentDate);
    }
}

File: src/backend/reservation/src/main/kotlin/org/taonaw/reservation/domain/model/cancellationFeeSetting/CancellationFeeSetting.kt
package org.taonaw.reservation.domain.model.cancellationFeeSetting

import org.taonaw.reservation.domain.model.reservation.UsageTime
import java.time.LocalDateTime

class CancellationFeeSetting(rates: List<CancellationFeeRate>) {
    val rates: List<CancellationFeeRate> = rates.sortedByDescending { it.daysAgo }

    fun chargedCancellationFee(usageTime: UsageTime, changedAt: LocalDateTime): Boolean {
        // TODO: 実装する
        return false
    }
}